framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceJavaParserStorage.java

```diff
@@ -997,6 +1024,44 @@ private static class ClassOrInterfaceAnnos {
     /** Collection of declared enum constants (empty if not an enum). */
     public Set<String> enumConstants = new HashSet<>(2);
 
+    /**
+     * Annotations on the declaration of the class (note that despite the name, these can also be
+     * type annotations).
+     */
+    private @MonotonicNonNull Set<AnnotationMirror> declarationAnnotations = null;
+
+    /**
+     * The Java Parser TypeDeclaration representing the class's declaration. Used for placing
+     * annotations inferred on the class declaration itself.
+     */
+    private @MonotonicNonNull TypeDeclaration<?> declaration = null;
+
+    /**
+     * Create a new ClassOrInterfaceAnnos
+     *
+     * @param javaParserNode the java parser node corresponding to the class declaration, which is
+     *     used for placing annotations on the class declaration
+     */
+    public ClassOrInterfaceAnnos(@Nullable TypeDeclaration<?> javaParserNode) {
+      if (javaParserNode != null) {
+        declaration = javaParserNode;
+      }
+    }
+
+    /**
+     * Adds {@code annotation} to the set of annotations on the declaration of this class.
+     *
+     * @param annotation an annotation (can be declaration or type)
+     * @return true if this is a new annotation for this class
+     */
+    public boolean addAnnotationToClassDeclaration(AnnotationMirror annotation) {
+      if (declarationAnnotations == null) {
+        declarationAnnotations = new HashSet<>();
```

Comment:
I am concerned that use of a HashSet might lead to nondeterministic WPI output.  The meaning will be the same, but it could affect diffs.

Reply:
This class already uses `HashSet` and `HashMap` to store other things whose order impacts diffs, e.g. declaration annotations on methods, the methods in a class, the fields in a class, etc. I think this is permitted, and they'll be sorted later (when the class is printed). 

(But, maybe all of those other uses are errors, too; if so, it might be better to fix them all at once separately.)